env/env_inspected.cc

```diff
@@ -0,0 +1,341 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+#ifndef ROCKSDB_LITE
+#include "rocksdb/env_inspected.h"
+
+namespace rocksdb {
+
+class InspectedSequentialFile : public SequentialFileWrapper {
+ public:
+  InspectedSequentialFile(std::unique_ptr<SequentialFile>&& target,
+                          FileSystemInspector* inspector)
+      : SequentialFileWrapper(target.get()),
+        owner_(std::move(target)),
+        inspector_(inspector) {}
+
+  Status Read(size_t n, Slice* result, char* scratch) override {
+    assert(inspector_);
+    Status s;
+    size_t offset = 0;
+    size_t allowed = 0;
+    while (offset < n) {
+      allowed = inspector_->Read(n - offset);
+      if (allowed > 0) {
+        s = SequentialFileWrapper::Read(allowed, result, scratch + offset);
+        if (!s.ok()) {
+          break;
+        }
+        size_t actual_read = result->size();
+        if (result->data() != scratch + offset) {
+          memcpy(scratch + offset, result->data(), actual_read);
+        }
+        offset += actual_read;
+        if (actual_read < allowed) {
+          break;
+        }
+      } else {
+        s = Status::IOError("Failed file system inspection");
```

Comment:
By design the inspector shouldn't return non-positive value at all. But simple assertion also seems too positive, thus the fail. I think retrying will make the code unnecessarily messy.

Reply:
I prefer to allow inspector to return 0, and the only error case being the inspector return error. The code is not more messy, since it only need to stay in the while-loop when the #bytes to read/write is not fullfill.

Reply:
You mean adding a status return for the inspector? because right now it can't return error. @yiwu-arbug

Reply:
I didn't realize the inspector doesn't return a status, but I think we should do that.